tsl/src/nodes/decompress_chunk/exec.c

```diff
@@ -384,80 +577,214 @@ initialize_batch(DecompressChunkState *state, TupleTableSlot *compressed_slot,
+	 * Reset expression memory context to clean out any cruft from
+	 * previous batch. Our batches are 1000 rows max, and this memory
+	 * context is used by ExecProject and ExecQual, which shouldn't
+	 * leak too much. So we only do this per batch and not per tuple to
+	 * save some CPU.
+	 */
+	econtext->ecxt_scantuple = decompressed_slot_scan;
+	ResetExprContext(econtext);
+
+	if (node->ss.ps.qual && !ExecQual(node->ss.ps.qual, econtext))
+	{
+		InstrCountFiltered1(node, 1);
+		return false;
+	}
+
+	if (node->ss.ps.ps_ProjInfo)
+	{
+		TupleTableSlot *projected = ExecProject(node->ss.ps.ps_ProjInfo);
+		ExecCopySlot(decompressed_slot_projected, projected);
+	}
+
+	return true;
+}
+
 static TupleTableSlot *
 decompress_chunk_exec(CustomScanState *node)
 {
-	DecompressChunkState *state = (DecompressChunkState *) node;
-	ExprContext *econtext = node->ss.ps.ps_ExprContext;
+	DecompressChunkState *chunk_state = (DecompressChunkState *) node;
 
 	if (node->custom_ps == NIL)
 		return NULL;
 
-	while (true)
+	/* If the batch_merge_append flag is set, the compression order_by and the
+	 * query order_by do match. Therefore, we use a binary heap to decompress
+	 * and merge the tuples.
+	 */
+	if (chunk_state->batch_merge_append)
 	{
-		TupleTableSlot *decompressed_slot = decompress_chunk_create_tuple(state);
+		/* Create the heap on the first call. */
+		if (chunk_state->merge_heap == NULL)
+		{
+			/* Prepare the heap and the batch states */
+			chunk_state->merge_heap = binaryheap_allocate(INITIAL_BATCH_CAPACITY,
+														  decompress_binaryheap_compare_slots,
+														  chunk_state);
+			batch_states_create(chunk_state, INITIAL_BATCH_CAPACITY);
+
+			/* Open the first batch */
+			decompress_batch_open_next_batch_and_add_to_binaryheap(chunk_state);
+		}
+		else
+		{
+			/* Remove the tuple returned in the last iteration and refresh the heap.
+			 * This operation is delayed up to this point where the next tuple actually
+			 * needs to be decompressed.
+			 */
+			decompress_batch_merge_remove_top_tuple_and_decompress_next(chunk_state);
+		}
 
-		if (TupIsNull(decompressed_slot))
+		/* All tuples are decompressed and consumed */
+		if (binaryheap_empty(chunk_state->merge_heap))
 			return NULL;
 
-		econtext->ecxt_scantuple = decompressed_slot;
-
-		/* Reset expression memory context to clean out any cruft from
-		 * previous tuple. */
-		ResetExprContext(econtext);
-
-		if (node->ss.ps.qual && !ExecQual(node->ss.ps.qual, econtext))
+		/* If the next tuple is from the top batch, open the next baches batches until
+		 * the next batch contains a tuple that is larger than the top tuple from the
+		 * heap (i.e., the batch is not the top element of the heap). */
+		while (DatumGetInt32(binaryheap_first(chunk_state->merge_heap)) ==
+			   chunk_state->batch_with_top_value)
 		{
-			InstrCountFiltered1(node, 1);
-			ExecClearTuple(decompressed_slot);
-			continue;
+			decompress_batch_open_next_batch_and_add_to_binaryheap(chunk_state);
 		}
```

Comment:
I don't really understand this part -- what happens if we have only one batch? The loop condition should be always true in that case?

Reply:
If we have consumed all compressed batches, `most_recent_batch` will be set to `INVALID_BATCH_ID`. So, the loop terminates at this point.

Reply:
Yes, but if it's not consumed yet? E.g. the underlying compressed scan has exactly one batch, it's also the most recent one, and we're trying to get next tuple and hitting this loop.

Reply:
In every iteration, `decompress_batch_open_next_batch` is called. This function (1) opens a new batch and updates `most_recent_batch` or (2) when all batches are consumed, it sets `most_recent_batch` to `INVALID_BATCH_ID`. So, the loop condition becomes false at some point.